src/encoding.ts

```diff
@@ -0,0 +1,379 @@
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function normalizeEncodingArgs(data: any, alternateCodec: any): [ string, ByteBuffer ]{
+	data = String(data);
+
+	return [ data, <ByteBuffer> {
+		toString: function (codec?: any): string {
+			return (codec || alternateCodec).decode(this);
+		},
+		toJSON: function (): string {
+			return JSON.stringify(Array.prototype.slice.call(this));
+		},
+		length: data.length
+	}];
+}
+
+function validateDecodingArgs(data: any) {
+	if (data == null) {
+		return false;
+	}
+
+	if (data.length == null) {
+		throw new TypeError('Argument must have a length property');
+	}
+
+	return true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [0, Infinity], checkSurrogate?: boolean): string {
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
+		throw Error('Surrogate is not a scalar value');
+	}
+
+	let encoded = '';
+
+	if (codePoint > 0xFFFF) {
+		codePoint = codePoint - 0x010000;
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+export interface ByteBuffer {
+	[index: number]: number;
+	toString?: (codec?: any) => string;
+	toJSON?: () => string;
+	length: number;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string (or string coerced object)
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export class Ascii {
```

Comment:
`Is there a point to these being classes rather than objects with 2 functions?`
There isn't really a great reason for these being classes over plain old objects. I opted for classes though because they can be extended. So, if someone were to want to subclass any of these they would be able to. I know thats not an amazing argument but when comparing the two options a class based approach is the more flexible of the two. That said I am not married to it and could easily be convinced that an object based approach is better

`Should there be an interface that these implement/conform to?`
There isn't an interface because all of the methods are static.